### PR TITLE
Prometheus Scaler Add custom headers and custom auth support

### LINKED ISSUE #4206: prometheus scaler: support list of custom headers and custom auth
### Proposal

Provide support for custom headers to be included in each queries in Prometheus scalar.

New prometheus scaler configuration being the `customHeaders` metadata:

Sample:
```yaml
kind: ScaledObject
metadata:
  name: prometheus-scaledobject
spec:
  triggers:
  - type: prometheus
    metadata:
      customHeaders: X-Client-Id=cid,X-Tenant-Id=tid,X-Organization-Id=oid
      authModes: "custom"
    authenticationRef:
      name: keda-prom-creds
---
kind: Secret
metadata:
  name: keda-prom-secret
data:
  customAuthHeader: "X-AUTH-TOKEN"
  customAuthValue: "auth-token"
---
kind: TriggerAuthentication
metadata:
  name: keda-prom-creds
spec:
  secretTargetRef:
    - parameter: customAuthHeader
      name: keda-prom-secret
      key: customAuthHeader
    - parameter: customAuthValue
      name: keda-prom-secret
      key: customAuthValue
```

### Use-Case

Custom headers can be helpful for various reasons:
- identification of tenant or organization in cloud environment
- identifier for the client or customer

Custom auth header can be useful for authentication and authorization purposes with any custom header name and value.

### Is this a feature you are interested in implementing yourself?

Yes

### Anything else?

Apart from authentication part, I see that only `x-Scope-OrgID` header is available, which is specific to querying multi tenant Cortex or Mimir.

That could be supported via `customHeaders` metadata as well.

### PR DESCRIPTION
**Pull request by prashant-shahi**
### Checklist

- [X] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [X] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [X] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to #4206

Relates to https://github.com/kedacore/keda-docs/pull/1064

### CODE REVIEW

File: pkg/scalers/prometheus_scaler.go [hunk 1]
```diff
@@ -160,6 +162,13 @@ func parsePrometheusMetadata(config *ScalerConfig) (meta *prometheusMetadata, er
 		meta.cortexOrgID = val
 	}
 
+	if val, ok := config.TriggerMetadata[promCustomHeaders]; ok && val != "" {
+		meta.customHeaders, err = kedautil.ParseStringList(val)
```
[the remark this answers is not in the document]
Comment (prashant-shahi) [reply pkg/scalers/prometheus_scaler.go]: nice catch!

File: CHANGELOG.md [hunk 2]
```diff
@@ -66,6 +66,7 @@ Here is an overview of all new **experimental** features:
 - **RabbitMQ Scaler**:  Add TLS support ([#967](https://github.com/kedacore/keda/issues/967))
 - **Redis Scalers**: Add support to Redis 7 ([#4052](https://github.com/kedacore/keda/issues/4052))
 - **Selenium Grid Scaler**: Add 'platformName' to selenium-grid scaler metadata structure ([#4038](https://github.com/kedacore/keda/issues/4038))
+- **Prometheus Scaler**: Add custom headers support ([#4037](https://github.com/kedacore/keda/issues/4208))
```
Comment (JorTurFer) [inline CHANGELOG.md]: Cloud you place this alphabetically sorted? (General is always the first things and then alphabetically sorted)

**Review by JorTurFer — COMMENTED** [review]
Thanks for the contribution! Looking good but I have one question:
Will these headers contain sensitive data? I mean, in the examples you have used "token". If these headers contain sensitive data, they should be read from TriggerAuthentication instead of metadata. We should avoid sensitive data as part of trigger metadata.
From the code pov, you should look for them `config.AuthParams` map instead of `config.TriggerMetadata`

File: CHANGELOG.md (hunk 2 above)
Comment (prashant-shahi) [reply CHANGELOG.md]: @JorTurFer updated

**Comment by prashant-shahi** [comment]
@JorTurFer yes, that makes sense for `custom` authentication header.

I have included changes related to the same and updated linked documentation PR as well.

**Review by JorTurFer — APPROVED** [review]
LGTM! Thanks for the improvement :) 
wdyt @zroubalik ?

**Review by zroubalik — COMMENTED** [review]
Generally looking good, what will happen if user specify both `cortexOrgID` option and a custom header where he specify `X-Scope-OrgID` header for Cortex but with a different value. Which one will be used and valid?

**Comment by prashant-shahi** [comment]
> Generally looking good, what will happen if user specify both `cortexOrgID` option and a custom header where he specify `X-Scope-OrgID` header for Cortex but with a different value. Which one will be used and valid?

@zroubalik Currently, the value from custom header will be used over `cortexOrgID`, when `X-Scope-OrgID` header is specified. However, I do not think one would be more valid than the other when both are passed.

Ideally, `cortexOrgID` can be depreciated after this PR, since the same can be fulfilled using custom headers.
We can mention in documentation to pass `cortexOrgID` using `X-Scope-OrgID` custom header.

**Comment by prashant-shahi** [comment]
@zroubalik Do let me know if there is any changes required for making the PR good to merge.

**Review by zroubalik — COMMENTED** [review]
@prashant-shahi the PR is looking good, I agree that we should deprecate the cortex parameter. Could you please add log message here if this parameter is specified, that custom headers should be used instead?

**Comment by zroubalik** [comment]
@prashant-shahi please also add note to Deprecated section in Changelog about this change.

**Comment by prashant-shahi** [comment]
@zroubalik `cortexOrgID` parameter has been deprecated and relevant files are updated.

**Review by zroubalik — APPROVED** [review]
LGTM, great job @prashant-shahi !


@tomkerkhove is there anything else needed to be done wrt deprecation? Could you please double check? 🙇‍♂️

**Comment by prashant-shahi** [comment]
@zroubalik Thank you! :)

I have also included checks for duplicate keys in custom headers as well as given higher precedence to authentication headers in case of collision with custom header.

**Comment by tomkerkhove** [comment]
> LGTM, great job @prashant-shahi !
> 
> @tomkerkhove is there anything else needed to be done wrt deprecation? Could you please double check? 🙇‍♂️

We're not there yet, we need to:

- Create GH discussion
- Create issue
- Document it with clear timelines

But we can do that after merging.

https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md#introducing-new-deprecations

I also noticed doc entry is not in our policy today so I have opened https://github.com/kedacore/governance/pull/95

**Comment by prashant-shahi** [comment]
> I also noticed doc entry is not in our policy today so I have opened https://github.com/kedacore/governance/pull/95

That's great! I was not sure about format for deprecation, so I went along with what I had in mind.

**Comment by tomkerkhove** [comment]
No worries!

**Pull request merged**